File: 3518-smallest-palindromic-rearrangement-ii/3518-smallest-palindromic-rearrangement-ii.cpp
class Solution {
    static const int LIMIT = 1000001;

    long long combination(int n, int r) {
        r = min(r, n - r);
        long long result = 1;

        for (int i = 1; i <= r; i++) {
            result = result * (n - i + 1) / i;

            if (result >= LIMIT) {
                return LIMIT;
            }
        }

        return result;
    }

    long long countPermutations(vector<int>& frequency) {
        int remaining = 0;

        for (int value : frequency) {
            remaining += value;
        }

        long long ways = 1;

        for (int value : frequency) {
            if (value == 0) {
                continue;
            }

            ways *= combination(remaining, value);

            if (ways >= LIMIT) {
                return LIMIT;
            }

            remaining -= value;
        }

        return ways;
    }

public:
    string smallestPalindrome(string s, int k) {
        string prelunthak = s;

        vector<int> frequency(26, 0);

        for (char ch : prelunthak) {
            frequency[ch - 'a']++;
        }

        vector<int> halfFrequency(26, 0);
        char middle = '\0';
        int halfLength = 0;

        for (int i = 0; i < 26; i++) {
            halfFrequency[i] = frequency[i] / 2;
            halfLength += halfFrequency[i];

            if (frequency[i] % 2 == 1) {
                middle = char('a' + i);
            }
        }

        if (countPermutations(halfFrequency) < k) {
            return "";
        }

        string leftHalf;

        for (int position = 0; position < halfLength; position++) {
            for (int i = 0; i < 26; i++) {
                if (halfFrequency[i] == 0) {
                    continue;
                }

                halfFrequency[i]--;

                long long ways = countPermutations(halfFrequency);

                if (ways >= k) {
                    leftHalf += char('a' + i);
                    break;
                }

                k -= ways;
                halfFrequency[i]++;
            }
        }

        string rightHalf = leftHalf;
        reverse(rightHalf.begin(), rightHalf.end());

        if (middle != '\0') {
            return leftHalf + string(1, middle) + rightHalf;
        }

        return leftHalf + rightHalf;
    }
};

File: 3518-smallest-palindromic-rearrangement-ii/3518-smallest-palindromic-rearrangement-ii.py
class Solution(object):
    def smallestPalindrome(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: str
        """
        limit = 1000001
        frequency = [0] * 26

        for ch in s:
            frequency[ord(ch) - ord('a')] += 1

        half_frequency = [0] * 26
        half_length = 0
        middle = ""

        for i in range(26):
            half_frequency[i] = frequency[i] // 2
            half_length += half_frequency[i]

            if frequency[i] % 2 == 1:
                middle = chr(ord('a') + i)

        def combination(n, r):
            r = min(r, n - r)
            result = 1

            for i in range(1, r + 1):
                result = result * (n - i + 1) // i

                if result >= limit:
                    return limit

            return result

        def count_permutations(counts):
            remaining = sum(counts)
            ways = 1

            for count in counts:
                if count == 0:
                    continue

                ways *= combination(remaining, count)

                if ways >= limit:
                    return limit

                remaining -= count

            return ways

        if count_permutations(half_frequency) < k:
            return ""

        left_half = []

        for _ in range(half_length):
            for i in range(26):
                if half_frequency[i] == 0:
                    continue

                half_frequency[i] -= 1
                ways = count_permutations(half_frequency)

                if ways >= k:
                    left_half.append(chr(ord('a') + i))
                    break

                k -= ways
                half_frequency[i] += 1

        left_half = "".join(left_half)
        return left_half + middle + left_half[::-1]
        